new-app should chase a defaulted "latest" tag to the stable ref

smarterclayton opened the pull request:
If an image stream has "latest" with an ImageStreamTag pointing to
another tag in the stream, and a user requests that stream via oc
new-app <stream>, new-app should create the app pointing to the
destination tag instead of "latest".  I.e.:

    # is mysql has three tags, latest->5.6, 5.6, and 5.6
    oc new-app mysql
    # create app pointing to mysql:5.6

Guarantees the admin can provide stable tags and the user ends up using a stable image stream.

smarterclayton commented:
@danmcp the last thing we needed to add from our original "openshift next" design.

smarterclayton commented:
@bparees fixed an issue in template search, a suggested name bug (for where we were suggesting the wrong app name on an existing image stream), and implemented the ref chase.

smarterclayton commented:
[test]

bparees commented on pkg/generate/app/componentresolvers.go:
```diff
@@ -39,41 +41,36 @@ type PerfectMatchWeightedResolver []WeightedResolver
 				if resolver.Weight != 0.0 {
 					m.Score = resolver.Weight * m.Score
 				}
 				imperfect = append(imperfect, m)
```
this name is confusing.  they are not necessarily imperfect matches, you could have a bunch of exact(perfect) matches in this array.

smarterclayton replied:
I may rename to "candidates"

On Thu, Jan 21, 2016 at 10:28 AM, Ben Parees <notifications@github.com>
wrote:

> In pkg/generate/app/componentresolvers.go
> <https://github.com/openshift/origin/pull/6670#discussion_r50414817>:
>
> >  			}
> > -			imperfect = append(imperfect, exact)
> > -		case len(inexact) > 0:
> > -			sort.Sort(ScoredComponentMatches(inexact))
> > -			if inexact[0].Score == 0.0 && (len(inexact) == 1 || inexact[1].Score != 0.0) {
> > -				return inexact[0], nil
> > -			}
> > -			for _, m := range inexact {
> > +			for _, m := range matches {
> >  				if resolver.Weight != 0.0 {
> >  					m.Score = resolver.Weight * m.Score
> >  				}
> >  				imperfect = append(imperfect, m)
>
> this name is confusing. they are not necessarily imperfect matches, you
> could have a bunch of exact(perfect) matches in this array.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670/files#r50414817>.
>

bparees replied:
+1

bparees commented on pkg/generate/app/imagestreamlookup.go:
```diff
@@ -33,9 +33,11 @@ func (r ImageStreamSearcher) Search(terms ...string) (ComponentMatches, error) {
 		if len(ref.Namespace) != 0 {
 			namespaces = []string{ref.Namespace}
 		}
+		followTag := false
```
comment this. i have no idea what a "followTag" is.

bparees commented on pkg/generate/app/imagestreamlookup.go:
```diff
@@ -33,9 +33,11 @@ func (r ImageStreamSearcher) Search(terms ...string) (ComponentMatches, error) {
 		searchTag := ref.Tag
 		if len(searchTag) == 0 {
 			searchTag = imageapi.DefaultImageTag
+			followTag = true
```
comment this.  on what basis did we decide to set followTag to true?  what user intent are we trying to execute on here?  ("user didn't specify a specific tag, therefore we are going to try to match to....")

bparees commented on pkg/generate/app/imagestreamlookup.go:
```diff
@@ -56,10 +58,20 @@ func (r ImageStreamSearcher) Search(terms ...string) (ComponentMatches, error) {
 					imageref, _ := imageapi.ParseDockerImageReference(term)
 					imageref.Name = stream.Name
+
+					localSearchTag := searchTag
```
why is this variable needed?  the original searchTag value is never referenced again after you assign localSearchTag below, so you could have just continued using searchTag.

smarterclayton replied:
Clarity of intent - localSearchTag is not the same thing as a searchTag,
can be a better name though.

On Thu, Jan 21, 2016 at 11:45 AM, Ben Parees <notifications@github.com>
wrote:

> In pkg/generate/app/imagestreamlookup.go
> <https://github.com/openshift/origin/pull/6670#discussion_r50426932>:
>
> > @@ -56,10 +58,20 @@ func (r ImageStreamSearcher) Search(terms ...string) (ComponentMatches, error) {
> >  				} else {
> >  					imageref, _ := imageapi.ParseDockerImageReference(term)
> >  					imageref.Name = stream.Name
> > +
> > +					localSearchTag := searchTag
>
> why is this variable needed? the original searchTag value is never
> referenced again after you assign localSearchTag below, so you could have
> just continued using searchTag.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670/files#r50426932>.
>

bparees replied:
near as i can tell the intent is "this is either the original searchTag provided by the user, or it is the Tag which is followed by the :latest tag", which seems hard to convey in a name, so maybe a comment would be better.

bparees commented on pkg/generate/app/imagestreamlookup.go:
```diff
@@ -56,10 +58,20 @@ func (r ImageStreamSearcher) Search(terms ...string) (ComponentMatches, error) {
 					imageref.Name = stream.Name
+
+					localSearchTag := searchTag
+					if specTag, ok := stream.Spec.Tags[searchTag]; ok && followTag {
```
// "if the user didn't specify a specific tag (so we're using latest), and the "latest" imagestreamtag is actually just 
// following another imagestreamtag, then use the actual imagestreamtag that "latest" is following as the match 
// result instead of the "latest"imagestream tag".  Also only do this if the name IST that "latest" is following does 
// not contain a ":" because....???

at least I think that's what this is doing, i'd like to save the next person who hits this code the trouble of trying to figure it out.

bparees commented on pkg/generate/app/templatelookup.go:
```diff
@@ -57,9 +61,13 @@ func (r TemplateSearcher) Search(terms ...string) (ComponentMatches, error) {
 			}
 		}
+		if exact {
+			break
```
shouldn't this be checking StopOnMatch?

smarterclayton replied:
exact is never set unless StopOnMatch is set.

On Thu, Jan 21, 2016 at 11:55 AM, Ben Parees <notifications@github.com>
wrote:

> In pkg/generate/app/templatelookup.go
> <https://github.com/openshift/origin/pull/6670#discussion_r50428380>:
>
> >  				}
> >  			}
> >  		}
> > +		if exact {
> > +			break
>
> shouldn't this be checking StopOnMatch?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670/files#r50428380>.
>

bparees replied:
I don't see any logic that's even using StopOnMatch in this PR.  what am i missing?

smarterclayton replied:
Nothing, it was a bug

bparees commented on pkg/generate/app/templatelookup.go:
```diff
@@ -21,34 +21,38 @@ type TemplateSearcher struct {
 	Client                    client.TemplatesNamespacer
 	TemplateConfigsNamespacer client.TemplateConfigsNamespacer
 	Namespaces                []string
+	StopOnMatch               bool
```
StopOnMatch or StopOnExactMatch?

if it's the latter, rename it or comment it.

bparees commented on pkg/generate/app/templatelookup.go:
```diff
@@ -57,9 +61,13 @@ func (r TemplateSearcher) Search(terms ...string) (ComponentMatches, error) {
 						Score:       score,
 						Template:    template,
 					})
+					break
```
i'm confused by this break.  what if the term you "matched" an item to was not an exact match?  don't you want to search on the other terms to see if there is a better match?  (i'm not clear under what conditions we'd actually have multiple terms either, but...)

smarterclayton replied:
If we had multiple terms we'd stop at the first level that had matches.
Not on the first match.

On Thu, Jan 21, 2016 at 11:58 AM, Ben Parees <notifications@github.com>
wrote:

> In pkg/generate/app/templatelookup.go
> <https://github.com/openshift/origin/pull/6670#discussion_r50428974>:
>
> > @@ -57,9 +61,13 @@ func (r TemplateSearcher) Search(terms ...string) (ComponentMatches, error) {
> >  						Score:       score,
> >  						Template:    template,
> >  					})
> > +					break
>
> i'm confused by this break. what if the term you "matched" an item to was
> not an exact match? don't you want to search on the other terms to see if
> there is a better match? (i'm not clear under what conditions we'd actually
> have multiple terms either, but...)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670/files#r50428974>.
>

bparees replied:
is that what you really want?  I guess i don't understand what the "levels" (multiple terms) are here...  but i would think if term1 had a lousy match, you'd still want to check to see if term2 had a good/better match.

bparees commented on pkg/generate/app/templatelookup.go:
```diff
@@ -92,7 +100,8 @@ func (r *TemplateFileSearcher) Search(terms ...string) (ComponentMatches, error)
 		if err != nil {
-			return nil, err
+			glog.V(5).Infof("Error from template search: %v", err)
+			continue
```
not aborting on an error seems like a mistake. or at least printing it as a warning.  Doesn't this mean if the template file the user intended to use has a typo, you're going to silently skip over that file?

smarterclayton replied:
Hrm, maybe.  I'm not happy with the error handling here.

On Thu, Jan 21, 2016 at 12:02 PM, Ben Parees <notifications@github.com>
wrote:

> In pkg/generate/app/templatelookup.go
> <https://github.com/openshift/origin/pull/6670#discussion_r50429491>:
>
> > @@ -92,7 +100,8 @@ func (r *TemplateFileSearcher) Search(terms ...string) (ComponentMatches, error)
> >  			Object()
> >
> >  		if err != nil {
> > -			return nil, err
> > +			glog.V(5).Infof("Error from template search: %v", err)
> > +			continue
>
> not aborting on an error seems like a mistake. or at least printing it as
> a warning. Doesn't this mean if the template file the user intended to use
> has a typo, you're going to silently skip over that file?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670/files#r50429491>.
>

bparees commented:
@smarterclayton review complete.

smarterclayton commented:
The last commit is a doozy but does greatly simplify the handling of errors.  I'm not convinced our error handling has enough tests, so I need to go through and add more multiple failure conditions and validate them.

smarterclayton commented:
Needs two fixes:

* Changes to returning errors means that DockerHub v2 fails out on an image that doesn't exist (because docker hub doesn't distinguish between this image exists and you don't have auth), so we'll need something new there
* Multiple error output needs to be thought through more - we probably want to coalesce errors of types into chunks.

bparees commented on pkg/generate/app/componentresolvers.go:
```diff
@@ -38,51 +41,47 @@ type PerfectMatchWeightedResolver []WeightedResolver
-					m.Score = resolver.Weight * m.Score
-				}
-				imperfect = append(imperfect, m)
+				m.Score = resolver.Weight * m.Score
```
though it doesn't really matter, this would be more readable if you used groupWeight instead of resolver.Weight here and above.

bparees commented on pkg/generate/app/componentresolvers.go:
```diff
@@ -90,58 +89,22 @@ func (r PerfectMatchWeightedResolver) Resolve(value string) (*ComponentMatch, er
 			}
-		}
-		return nil, ErrNoMatch{value: value}
+		}*/
```
is this getting uncommented prior to merge?  or hoping for @rhcarvalho's changes?

smarterclayton replied:
It's not needed any more - you see the error along with no match

smarterclayton replied:
```
$ oc new-app foo.json
error: unable to load template file "foo.json": invalid character '}' after object key
error: no match for "foo.json"

The 'oc new-app' command will match arguments to the following types:

  1. Images tagged into image streams in the current project or the 'openshift' project
     - if you don't specify a tag, we'll add ':latest'
  2. Images in the Docker Hub, on remote registries, or on the local Docker engine
  3. Templates in the current project or the 'openshift' project
  4. Git repository URLs or local paths that point to Git repositories

--allow-missing-images can be used to point to an image that does not exist yet.

See 'oc new-app -h' for examples.
```

Will add a test explicitly for this.

bparees commented on pkg/generate/app/componentresolvers.go:
```diff
@@ -197,21 +154,18 @@ type HighestUniqueScoreResolver struct {
 // Resolve resolves as the highest scored match returned by the Searcher, and
 // guarantees the match is unique (the only match with that given score)
 func (r HighestUniqueScoreResolver) Resolve(value string) (*ComponentMatch, error) {
```
do we really have use cases for all these resolvers?  (first, highest, highest unique)
seems like "highest score" by itself ought to be sufficient.

smarterclayton replied:
Maybe not, but I don't want to introduce even more churn in this PR.  They're in the resolver chain today for the other types

bparees commented on pkg/generate/app/dockerimagelookup.go:
```diff
@@ -41,13 +40,18 @@ type DockerClientSearcher struct {
 	componentMatches := ComponentMatches{}
 	errs := []error{}
 	for _, term := range terms {
+		if term == "__dockerimage_fail" {
```
debugging?

smarterclayton replied:
More like something concrete we can use in test cases.  I occasionally like being able to make something fail concretely.

bparees commented on pkg/generate/app/imagestreamlookup.go:
```diff
@@ -18,35 +18,45 @@ type ImageStreamSearcher struct {
+			glog.V(2).Infof("image streams must be of the form [<namespace>/]<name>[:<tag>|@<digest>], term %q did not qualify", term)
+			continue
+		}
+		if term == "__imagestream_fail" {
```
debugging?

bparees commented on pkg/generate/app/imagestreamlookup.go:
```diff
@@ -56,28 +66,41 @@ func (r ImageStreamSearcher) Search(terms ...string) (ComponentMatches, error) {
+				// When an image stream contains a tag that references another local tag, and the user has not
+				// provided a tag themselves (i.e. they asked for mysql and we defaulted to mysql:latest), walk
+				// the chain of references to the end. This ensures that applications can default to using a "stable"
+				// branch by giving the control over version to the image stream author.
```
this comment brings me great joy.

rhcarvalho commented:
Maybe [new-app should chase a defaulted "latest" tag to the stable ref](https://github.com/smarterclayton/origin/commit/01432fe4edce8619c14454960a64222ccb207a39) and [Add a test for chasing references in new-app](https://github.com/smarterclayton/origin/commit/a10996f2bedb147174bf3d4b4fcf900fa4dcc5d2) should be squashed together?

rhcarvalho commented on pkg/generate/app/imagestreamlookup.go:
```diff
@@ -278,19 +301,32 @@ func (r *ImageStreamByAnnotationSearcher) annotationMatches(stream *imageapi.Ima
 		}
 		for i := range streams {
 			for _, term := range terms {
+				if term == "__imagestreamannotation_fail" {
```
debugging?

smarterclayton commented:
Ok, this is updated with logic to unify the errors.  I went through and tried to whitelist certain error conditions as "known not found".  The docker hub one is the hardest, because DockerHub gives us Unauthorized for "NotFound' in some cases, and that's not a great experience.

smarterclayton commented:
[test]

On Sun, Jan 31, 2016 at 7:00 PM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/test FAILURE (
> https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/626/)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670#issuecomment-177664013>.
>

smarterclayton commented:
[test]

On Sun, Jan 31, 2016 at 8:45 PM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/test FAILURE (
> https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/628/)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670#issuecomment-177692900>.
>

smarterclayton commented:
Stabs bash repeatedly in the neck and then goes to sleep.

On Sun, Jan 31, 2016 at 10:00 PM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/test FAILURE (
> https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/629/)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670#issuecomment-177727831>.
>

smarterclayton commented:
Zomg

smarterclayton commented:
[test]

smarterclayton commented:
I don't believe this fixed the import flakes

[test]

On Mon, Feb 1, 2016 at 9:22 PM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/test SUCCESS (
> https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/678/)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670#issuecomment-178317777>.
>

smarterclayton commented:
[test]

smarterclayton commented:
ZOMG

On Mon, Feb 1, 2016 at 10:51 PM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/test SUCCESS (
> https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/683/)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670#issuecomment-178353813>.
>

smarterclayton commented:
Four non flaky runs in a row - let's try to close this review out tomorrow and maybe we'll be better off

smarterclayton commented:
[test]

rhcarvalho commented on test/cmd/newapp.sh:
```diff
@@ -9,10 +9,18 @@ source "${OS_ROOT}/hack/util.sh"
+  set +e
+  oc delete all,templates --all
+  exit 0
+) &>/dev/null
```
Is this supposed to be in the final patch?
Why do we delete resources prior to running the tests?

smarterclayton replied:
So we can rerun them as an end user - the original point of all the
individual scripts is that you can re run them over and over.

On Tue, Feb 2, 2016 at 3:19 AM, Rodolfo Carvalho <notifications@github.com>
wrote:

> In test/cmd/newapp.sh
> <https://github.com/openshift/origin/pull/6670#discussion_r51536701>:
>
> > @@ -9,10 +9,18 @@ source "${OS_ROOT}/hack/util.sh"
> >  source "${OS_ROOT}/hack/cmd_util.sh"
> >  os::log::install_errexit
> >
> > +# Cleanup cluster resources created by this test
> > +(
> > +  set +e
> > +  oc delete all,templates --all
> > +  exit 0
> > +) &>/dev/null
>
> Is this supposed to be in the final patch?
> Why do we delete resources prior to running the tests?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670/files#r51536701>.
>

rhcarvalho replied:
The pattern I've seen before in these tests was to do clean ups only when necessary between test cases. The "test framework" automatically creates a project for each `test/cmd/*.sh` file and after the tests run it will remove the whole project.

So the way I've been running these tests locally during development is:

```console
$ hack/test-cmd.sh test/cmd/newapp.sh
```

Perhaps you're running them in a smarter way?

What I don't like in the way I run it is that it takes a way too long time to start a server in every test run, making developing those tests quite painful. I often comment out all other tests in a file.

@smarterclayton could you describe how you develop and run test/cmd tests locally? I'd happily add the knowledge to [HACKING.md](https://github.com/openshift/origin/blob/master/HACKING.md#integration-tests).

smarterclayton replied:
What I'm doing is starting up the master in another cmd shell (openshift
start master --loglevel=5) and then just running test/cmd/newapp.sh.  A few
tests need preconditions, but any test that needs preconditions should set
them up anyway (for test-cmd).

On Tue, Feb 2, 2016 at 11:00 AM, Rodolfo Carvalho <notifications@github.com>
wrote:

> In test/cmd/newapp.sh
> <https://github.com/openshift/origin/pull/6670#discussion_r51588259>:
>
> > @@ -9,10 +9,18 @@ source "${OS_ROOT}/hack/util.sh"
> >  source "${OS_ROOT}/hack/cmd_util.sh"
> >  os::log::install_errexit
> >
> > +# Cleanup cluster resources created by this test
> > +(
> > +  set +e
> > +  oc delete all,templates --all
> > +  exit 0
> > +) &>/dev/null
>
> The pattern I've seen before in these tests was to do clean ups only when
> necessary between test cases. The "test framework" automatically creates a
> project for each test/cmd/*.sh file and after the tests run it will
> remove the whole project.
>
> So the way I've been running these tests locally during development is:
>
> $ hack/test-cmd.sh test/cmd/newapp.sh
>
> Perhaps you're running them in a smarter way?
>
> What I don't like in the way I run it is that it takes a way too long time
> to start a server in every test run, making developing those tests quite
> painful. I often comment out all other tests in a file.
>
> @smarterclayton <https://github.com/smarterclayton> could you describe
> how you develop and run test/cmd tests locally? I'd happily add the
> knowledge to HACKING.md
> <https://github.com/openshift/origin/blob/master/HACKING.md#integration-tests>
> .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670/files#r51588259>.
>

rhcarvalho replied:
I could document that during development one could run `test/cmd/*.sh` files to test against a running server.

However, this also means that running a test might accidentally delete resources you had in the server.

Could we protect this clean up somehow? Say, only delete resources if an environment variable `OPENSHIFT_TEST_CMD_FORCE_CLEANUP` is set.
Then, by default, running a test would not delete resources in whatever namespace/project you happen to be.

smarterclayton replied:
That's possible.  But you could also accidentally delete things by running
oc delete all --all in normal use, or run rm -rf *.  At a certain level,
running tests accepts a risk.

On Tue, Feb 2, 2016 at 12:04 PM, Rodolfo Carvalho <notifications@github.com>
wrote:

> In test/cmd/newapp.sh
> <https://github.com/openshift/origin/pull/6670#discussion_r51598881>:
>
> > @@ -9,10 +9,18 @@ source "${OS_ROOT}/hack/util.sh"
> >  source "${OS_ROOT}/hack/cmd_util.sh"
> >  os::log::install_errexit
> >
> > +# Cleanup cluster resources created by this test
> > +(
> > +  set +e
> > +  oc delete all,templates --all
> > +  exit 0
> > +) &>/dev/null
>
> I could document that during development one could run test/cmd/*.sh
> files to test against a running server.
>
> However, this also means that running a test might accidentally delete
> resources you had in the server.
>
> Could we protect this clean up somehow? Say, only delete resources if an
> environment variable OPENSHIFT_TEST_CMD_FORCE_CLEANUP is set.
> Then, by default, running a test would not delete resources in whatever
> namespace/project you happen to be.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670/files#r51598881>.
>

smarterclayton replied:
Pushed an update to move the test json into fixtures.

On Tue, Feb 2, 2016 at 12:06 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> That's possible.  But you could also accidentally delete things by running
> oc delete all --all in normal use, or run rm -rf *.  At a certain level,
> running tests accepts a risk.
>
> On Tue, Feb 2, 2016 at 12:04 PM, Rodolfo Carvalho <
> notifications@github.com> wrote:
>
>> In test/cmd/newapp.sh
>> <https://github.com/openshift/origin/pull/6670#discussion_r51598881>:
>>
>> > @@ -9,10 +9,18 @@ source "${OS_ROOT}/hack/util.sh"
>> >  source "${OS_ROOT}/hack/cmd_util.sh"
>> >  os::log::install_errexit
>> >
>> > +# Cleanup cluster resources created by this test
>> > +(
>> > +  set +e
>> > +  oc delete all,templates --all
>> > +  exit 0
>> > +) &>/dev/null
>>
>> I could document that during development one could run test/cmd/*.sh
>> files to test against a running server.
>>
>> However, this also means that running a test might accidentally delete
>> resources you had in the server.
>>
>> Could we protect this clean up somehow? Say, only delete resources if an
>> environment variable OPENSHIFT_TEST_CMD_FORCE_CLEANUP is set.
>> Then, by default, running a test would not delete resources in whatever
>> namespace/project you happen to be.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/openshift/origin/pull/6670/files#r51598881>.
>>
>
>

rhcarvalho replied:
> But you could also accidentally delete things by running
> oc delete all --all in normal use,

I've hit this while working on the project :smile: . The catch here is that whenever the development server is restarted, it sends you back to the `default` namespace. https://github.com/openshift/origin/issues/2516

I "learned" to always include a "-n" argument in my `oc delete` calls.


> or run rm -rf *

I wouldn't like a project that deletes my files when I run the test suite. This reminds of the epic [bumblebee](https://github.com/MrMEEE/bumblebee-Old-and-abbandoned/commit/a047be85247755cdbe0acce6f1dafc8beb84f2ac#diff-3fbb47e318cd8802bd325e7da9aaabe8L351).


I wish there were no "risks" involved when running tests.
IMHO, it's easier to tolerate test suites that leaves behind temporary files (as long as it doesn't make your disk bleed) than having files changed or removed.

smarterclayton replied:
If we didn't put the registry or kubernetes service into the default
namespace, it would be safer.  That's been under discussion for awhile.

On Tue, Feb 2, 2016 at 12:20 PM, Rodolfo Carvalho <notifications@github.com>
wrote:

> In test/cmd/newapp.sh
> <https://github.com/openshift/origin/pull/6670#discussion_r51601425>:
>
> > @@ -9,10 +9,18 @@ source "${OS_ROOT}/hack/util.sh"
> >  source "${OS_ROOT}/hack/cmd_util.sh"
> >  os::log::install_errexit
> >
> > +# Cleanup cluster resources created by this test
> > +(
> > +  set +e
> > +  oc delete all,templates --all
> > +  exit 0
> > +) &>/dev/null
>
> But you could also accidentally delete things by running
> oc delete all --all in normal use,
>
> I've hit this while working on the project [image: :smile:] . The catch
> here is that whenever the development server is restarted, it sends you
> back to the default namespace. #2516
> <https://github.com/openshift/origin/issues/2516>
>
> I "learned" to always include a "-n" argument in my oc delete calls.
>
> or run rm -rf *
>
> I wouldn't like a project that deletes my files when I run the test suite.
> This reminds of the epic bumblebee
> <https://github.com/MrMEEE/bumblebee-Old-and-abbandoned/commit/a047be85247755cdbe0acce6f1dafc8beb84f2ac#diff-3fbb47e318cd8802bd325e7da9aaabe8L351>
> .
>
> I wish there were no "risks" involved when running tests.
> IMHO, it's easier to tolerate test suites that leaves behind temporary
> files (as long as it doesn't make your disk bleed) than having files
> changed or removed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670/files#r51601425>.
>

rhcarvalho commented on test/cmd/newapp.sh:
```diff
@@ -126,6 +158,10 @@ os::cmd::expect_success 'oc delete template ruby-helloworld-sample'
 os::cmd::expect_success 'oc delete all -l app=ruby'
+# check for error when template JSON file has errors
+jsonfile="${TMPDIR:-/tmp}/foo.json"
+echo '{"test"}' > "${jsonfile}"
```
I searched for `TMPDIR` in the project, but didn't find any place that would set it to anything other than `/tmp`.
I'm assuming this file will not be deleted by the test framework after run, is it to be considered a problem?

smarterclayton replied:
I don't expect it is - we simply have to have a file on disk somewhere.  I
may just move this to a fixture.

On Tue, Feb 2, 2016 at 3:41 AM, Rodolfo Carvalho <notifications@github.com>
wrote:

> In test/cmd/newapp.sh
> <https://github.com/openshift/origin/pull/6670#discussion_r51538538>:
>
> > @@ -126,6 +158,10 @@ os::cmd::expect_success 'oc delete template ruby-helloworld-sample'
> >  os::cmd::expect_success_and_text 'oc new-app mysql --name=db' 'db'
> >  os::cmd::expect_success 'oc new-app https://github.com/openshift/ruby-hello-world -l app=ruby'
> >  os::cmd::expect_success 'oc delete all -l app=ruby'
> > +# check for error when template JSON file has errors
> > +jsonfile="${TMPDIR:-/tmp}/foo.json"
> > +echo '{"test"}' > "${jsonfile}"
>
> I searched for TMPDIR in the project, but didn't find any place that
> would set it to anything other than /tmp.
> I'm assuming this file will not be deleted by the test framework after
> run, is it to be considered a problem?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670/files#r51538538>.
>

rhcarvalho commented on test/cmd/newapp.sh:
```diff
@@ -126,6 +158,10 @@ os::cmd::expect_success 'oc delete template ruby-helloworld-sample'
+# check for error when template JSON file has errors
+jsonfile="${TMPDIR:-/tmp}/foo.json"
+echo '{"test"}' > "${jsonfile}"
+os::cmd::expect_failure_and_text "oc new-app '${jsonfile}'" "error: unable to load template file \"${jsonfile}\": invalid character '}' after object key"
```
In practice, don't you think users will miss this error message when it is followed by a block of 10+ lines telling how new-app works?
I believe people may read the new-app instructions once, and in all subsequent runs our brains just pattern-matches the block of text and ignore it.

smarterclayton replied:
In practice we print:

$ commandline
<chunk of errors>
<chunk of text>

My concern about

$ commandline
<title of one error>
<chunk of text>
<chunk of errors>

is that the latter is also hard to parse.


On Tue, Feb 2, 2016 at 3:44 AM, Rodolfo Carvalho <notifications@github.com>
wrote:

> In test/cmd/newapp.sh
> <https://github.com/openshift/origin/pull/6670#discussion_r51538818>:
>
> > @@ -126,6 +158,10 @@ os::cmd::expect_success 'oc delete template ruby-helloworld-sample'
> >  os::cmd::expect_success_and_text 'oc new-app mysql --name=db' 'db'
> >  os::cmd::expect_success 'oc new-app https://github.com/openshift/ruby-hello-world -l app=ruby'
> >  os::cmd::expect_success 'oc delete all -l app=ruby'
> > +# check for error when template JSON file has errors
> > +jsonfile="${TMPDIR:-/tmp}/foo.json"
> > +echo '{"test"}' > "${jsonfile}"
> > +os::cmd::expect_failure_and_text "oc new-app '${jsonfile}'" "error: unable to load template file \"${jsonfile}\": invalid character '}' after object key"
>
> In practice, don't you think users will miss this error message when it is
> followed by a block of 10+ lines telling how new-app works?
> I believe people may read the new-app instructions once, and in all
> subsequent runs our brains just pattern-matches the block of text and
> ignore it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670/files#r51538818>.
>

rhcarvalho replied:
If you have a file in the current directory named `template.json` and run `oc new-app template.json`, what's the likelihood you meant a Docker image called "template.json"?

I think in that case new-app should just say the file has invalid JSON (+ details from the parser), and print nothing else.

smarterclayton replied:
Hrm - that would mean the template file resolver has to have top weight,
which means a local file will *always* take precedence over any image from
anywhere.  So having a file called php in the local directory would prevent
you from being able to create php things.  I agree with the statement, but
I think we'd have to only look at things with .yaml and .json extensions to
give them that weight (php.json I can buy being weighted above everything
else, php i can't).

On Tue, Feb 2, 2016 at 11:04 AM, Rodolfo Carvalho <notifications@github.com>
wrote:

> In test/cmd/newapp.sh
> <https://github.com/openshift/origin/pull/6670#discussion_r51588896>:
>
> > @@ -126,6 +158,10 @@ os::cmd::expect_success 'oc delete template ruby-helloworld-sample'
> >  os::cmd::expect_success_and_text 'oc new-app mysql --name=db' 'db'
> >  os::cmd::expect_success 'oc new-app https://github.com/openshift/ruby-hello-world -l app=ruby'
> >  os::cmd::expect_success 'oc delete all -l app=ruby'
> > +# check for error when template JSON file has errors
> > +jsonfile="${TMPDIR:-/tmp}/foo.json"
> > +echo '{"test"}' > "${jsonfile}"
> > +os::cmd::expect_failure_and_text "oc new-app '${jsonfile}'" "error: unable to load template file \"${jsonfile}\": invalid character '}' after object key"
>
> If you have a file in the current directory named template.json and run oc
> new-app template.json, what's the likelihood you meant a Docker image
> called "template.json"?
>
> I think in that case new-app should just say the file has invalid JSON (+
> details from the parser), and print nothing else.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670/files#r51588896>.
>

rhcarvalho replied:
I agree with you @smarterclayton :+1:

smarterclayton commented:
Crap, new flake:

```

--- FAIL: TestImageStreamImportDockerHub-2 (15.00s)
	imageimporter_test.go:507: unexpected repository:
&api.RepositoryImportStatus{Status:unversioned.Status{TypeMeta:unversioned.TypeMeta{Kind:"",
APIVersion:""}, ListMeta:unversioned.ListMeta{SelfLink:"",
ResourceVersion:""}, Status:"Failure", Message:"Internal error
occurred: Get https://registry-1.docker.io/v2/: read tcp
52.72.23.195:443: use of closed network connection",
Reason:"InternalError",
Details:(*unversioned.StatusDetails)(0xc208609c00), Code:500},
Images:[]api.ImageImportStatus(nil), AdditionalTags:[]string(nil)}
	imageimporter_test.go:514: unexpected object: (*api.Image)(nil)
	imageimporter_test.go:518: unexpected object: (*api.Image)(nil)
	imageimporter_test.go:522: unexpected object: (*api.Image)(nil)
	imageimporter_test.go:526: unexpected object:
api.ImageImportStatus{Tag:"",
Status:unversioned.Status{TypeMeta:unversioned.TypeMeta{Kind:"",
APIVersion:""}, ListMeta:unversioned.ListMeta{SelfLink:"",
ResourceVersion:""}, Status:"Failure", Message:"Internal error
occurred: Get https://registry-1.docker.io/v2/: read tcp
52.72.23.195:443: use of closed network connection",
Reason:"InternalError",
Details:(*unversioned.StatusDetails)(0xc208609a80), Code:500},
Image:(*api.Image)(nil)}
FAIL

```

​
[test]

bparees commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -204,33 +206,46 @@ func handleBuildError(c *cobra.Command, err error, fullName string) error {
-		return err
-	}
+		groups.Add(
+			"no-matches",
```
i'm not seeing how this handling is different from the newapp.transformError() implementation?

smarterclayton replied:
3 suggestions vs 4

bparees replied:
ah right.

bparees commented on pkg/generate/app/componentresolvers.go:
```diff
@@ -197,21 +139,18 @@ type HighestUniqueScoreResolver struct {
 	default:
 		if matches[0].Score == matches[1].Score {
-			return nil, ErrMultipleMatches{value, matches}
+			return nil, ErrMultipleMatches{Value: value, Matches: matches, Errs: err}
```
should this only return(in the error object) the matches that have the same score, rather than all the matches?

smarterclayton replied:
yes

bparees commented on pkg/generate/app/dockerimagelookup.go:
```diff
@@ -41,13 +40,18 @@ type DockerClientSearcher struct {
 		ref, err := imageapi.ParseDockerImageReference(term)
 		if err != nil {
-			return nil, err
+			continue
```
continue?  shouldn't this at least aggregate the error?

smarterclayton replied:
passing a term that is not a valid docker image reference is not an error, it is simply "there is nothing that could possibly match this".  If we didn't parse the reference, or had partial matching, we could.

bparees replied:
as a user i feel like i'd want to be told that, rather than repeatedly being told "nothing matched your stupid typo"

bparees replied:
especially since they might go on to use the --allow-missing-images flag w/ their invalid argument.

smarterclayton replied:
But we show you the typo.  Returning an error should actually be an error
that applies to the term - a term that can't be an image isn't really an
error.  "../foo/bar/template.json" isn't an error

On Tue, Feb 2, 2016 at 3:21 PM, Ben Parees <notifications@github.com> wrote:

> In pkg/generate/app/dockerimagelookup.go
> <https://github.com/openshift/origin/pull/6670#discussion_r51628839>:
>
> >  		ref, err := imageapi.ParseDockerImageReference(term)
> >  		if err != nil {
> > -			return nil, err
> > +			continue
>
> as a user i feel like i'd want to be told that, rather than repeatedly
> being told "nothing matched your stupid typo"
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670/files#r51628839>.
>

bparees replied:
ok that makes sense, i wasn't considering the fact that this function is going to get called w/ args that were never intended to be matched to image specs.

bparees commented on pkg/generate/app/dockerimagelookup.go:
```diff
@@ -94,6 +95,10 @@ func (r DockerClientSearcher) Search(terms ...string) (ComponentMatches, error)
 		sort.Sort(termMatches)
 
+		if r.Client == nil {
+			continue
```
this doesn't look right...  r.Client was already used on line 76 (assuming no exact matches in the registry)

smarterclayton replied:
Will move it up above 70

bparees commented on pkg/generate/app/dockerimagelookup.go:
```diff
@@ -161,9 +164,14 @@ type ImageImportSearcher struct {
 }
 
-func (s ImageImportSearcher) Search(terms ...string) (ComponentMatches, error) {
+func (s ImageImportSearcher) Search(precise bool, terms ...string) (ComponentMatches, []error) {
```
Godoc

bparees commented on pkg/generate/app/errors.go:
```diff
@@ -4,77 +4,67 @@ import (
-func (e ErrNoMatch) UsageError(commandName string) string {
-	return fmt.Sprintf("%[3]s - does a Docker image with that name exist?", e.value, commandName, e.Error())
+func (e ErrNoMatch) Suggestion(commandName string) string {
+	return fmt.Sprintf("%[3]s - does a Docker image with that name exist?", e.Value, commandName, e.Error())
```
shouldn't this be suggesting --allow-missing-images?

smarterclayton replied:
newapp and newbuild are now responsible for that

bparees commented:
@smarterclayton i think that's all i've got.

smarterclayton commented:
Squashed and rebased

On Tue, Feb 2, 2016 at 3:50 PM, OpenShift Bot <notifications@github.com>
wrote:

> Evaluated for origin test up to 00b8aff
> <https://github.com/openshift/origin/commit/00b8aff11e89612fa29c157643b3a9ce94bece04>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670#issuecomment-178812966>.
>

bparees commented:
lgtm

smarterclayton commented:
[merge] and hope this fixes flakes

smarterclayton commented:
[test]

smarterclayton commented:
[test] - guess I just got really lucky.  oh well.

On Tue, Feb 2, 2016 at 5:55 PM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/test FAILURE (
> https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/749/)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6670#issuecomment-178873856>.
>
